lib/magma/server/update.rb

```diff
@@ -2,65 +2,179 @@
-      error(revision.errors) if !revision.valid?
+      @payload.add_records(model, model_revisions.map{ |rev| rev.to_payload(@user) })
     end
+
+    @loader.dispatch_record_set
+  rescue Magma::LoadFailed => m
+    log(m.complaints)
+    @errors.concat(m.complaints)
+  end
+
+  private
+
+  def is_file_attribute(model, attribute)
+    return model.attributes[attribute].is_a? Magma::FileAttribute
   end
 
-  def post_revisions
-    revisions.each do |revision|
-      begin
-        revision.post!
-      rescue Magma::LoadFailed => m
-        log m.complaints
-        @errors.concat m.complaints
-        next
+  def update_any_file_links
+    # Here, if there are any File attributes in the revisions,
+    #   we'll send a request to the Metis "copy" route, using the
+    #   Etna::Client.
+    # For ::blank or ::nil revisions, we remove the link in Metis.
+    # Assumes a conversion to Metis storage and no other
+    #   provider will be set.
+
+    # Etna::Client.new throws an exception if the host config includes protocol,
+    # i.e. https://metis-dev.etna-development.org will throw an exception
+    # raise Etna::BadRequest, 'Storage host should not include protocol' if
+    #   Magma.instance.config(:storage).fetch(:host).start_with? 'http'
+
+    copy_revisions = []
+
+    @revisions.each do |model, model_revisions|
+      model_revisions.each do |revision|
+
+        revision.attribute_names.select {
+          |attribute_name| is_file_attribute(revision.model, attribute_name)
+        }.each do |attribute_name|
+          # For the remove operations, have to figure
+          #   out the "current" link name in Metis,
+          #   using the stored Magma value.
+          if !revision[attribute_name] || revision[attribute_name] == '::blank'
+            retrieval = Magma::Retrieval.new(
+              revision.model,
+              [revision.record_name.to_s],
+              [attribute_name],
+              restrict: !@user.can_see_restricted?(@project_name)
+            )
+
+            retrieval.records.each do |record|
+              # Only add a link to remove on Metis if a copy already exists.
+              # If the attribute hasn't been set before, we'll ignore this.
+              if record[attribute_name]
+                # This Metis route isn't active yet, so don't do anything
+                # link_revisions.push({
+                #   source: "metis://#{@project_name}/magma/#{record[attribute_name][:path]}",
+                #   dest: nil
+                # })
+              end
+            end
+          elsif revision[attribute_name].start_with? 'metis://'
+            copy_revisions.push({
+              source: revision[attribute_name],
+              dest: "metis://#{@project_name}/magma/#{revision.to_loader[attribute_name][:filename]}"
+            })
+          end
+        end
       end
     end
+
+    if copy_revisions.length > 0 && Magma.instance.storage.instance_of?(Magma::Storage::Metis)
```

Comment:
This storage check is to make the merge into master easier, until we make a full transition to Metis storage. Otherwise Magma would break the instance we merged this in, when using AWS storage.

Can remove the check at a later point, after the full transition.